needy/caches/directorycache.py

```diff
@@ -0,0 +1,55 @@
+import tarfile
+import os
+from datetime import datetime, timedelta
+
+from ..cache import Cache
+
+
+class DirectoryCache(Cache):
+    def __init__(self, configuration):
+        Cache.__init__(self, configuration)
+
+    @staticmethod
+    def type():
+        return 'directory'
+
+    def description(self):
+        return self.__path
+
+    def archive(self, source, key):
+        destination_path = self.__archive_path(key)
+        parent_path = os.path.dirname(destination_path)
+        if not os.path.exists(parent_path):
+            os.makedirs(parent_path)
+        if os.path.exists(destination_path):
+            os.remove(destination_path)
+        tar = tarfile.open(destination_path, 'w:gz')
+        tar.add(source, arcname='.')
+        tar.close()
+        self.__garbage_collect()
+
+    def unarchive(self, destination, key):
+        tar = tarfile.open(self.__archive_path(key), 'r:gz')
+        tar.extractall(path=destination)
+        tar.close()
+        os.utime(self.__archive_path(key), None)
+
+    def exists(self, key):
+        return os.path.exists(self.__archive_path(key))
+
+    def __garbage_collect(self, min_date=datetime.now()+timedelta(weeks=-2)):
+        for root, directories, filenames in os.walk(self.__archive_path()):
+            for filename in filenames:
+                path = os.path.join(root, filename)
+                try:
+                    mtime = os.path.getmtime(path)
+                except OSError:
+                    mtime = 0
+                last_modified_date = datetime.fromtimestamp(mtime)
+                if last_modified_date < min_date:
+                    logging.debug('Removing older cache item {}'.format(path))
+                    os.remove(path)
```

Comment:
1) EFS looks awesome.

2) I'm not convinced we couldn't use S3 and still have correct behavior. We would have to make sure that after a key was declared available, it was (exists method would download the object if not present locally) and surely download/delete operations on s3 objects are atomic. This applies whether or not we garbage collect just for robustnes' sake.

3) For s3, I think we would want to configure expiration policies instead of managing that via needy. Needy might provide some facilities to create/configure a bucket, but otherwise it makes the most sense to me to let AWS handle expiration. Less flexibility for sure, easier to manage (cause we don't!).